skyve-ext/src/main/java/org/skyve/impl/util/WebStatsUtil.java

```diff
@@ -2,43 +2,155 @@
+
+		// Check if the IpInfo token has been set so as to get the country code
+		String ipInfoToken = UtilImpl.IP_INFO_TOKEN;
+		String countryCode = null;
+		if (ipInfoToken != null) {
+			final GeoIPService geoIPService = new GeoIPService();
+			Optional<String> countryCodeOptional = geoIPService.getCountryCodeForIP(userIPAddress);
+			if (countryCodeOptional.isPresent()) {
+				countryCode = countryCodeOptional.get();
+				BindUtil.set(loginRecord, COUNTRY, countryCode);
+			}
+		}
+		AbstractPersistence.get()
+				.save(loginRecordDocument, loginRecord);
+
+		// Get the last login record of the current user
+		DocumentQuery q = CORE.getPersistence()
+				.newDocumentQuery(loginRecordDocument);
+		q.getFilter()
+				.addEquals(AppConstants.USER_NAME_ATTRIBUTE_NAME, user.getName());
+		q.addBoundOrdering(LOGIN_DATE_TIME, SortDirection.descending);
+		AbstractPersistentBean previousLoginRecord = q.beanResult();
+
+		// Get email and name from the metadata user
+		Document userDocument = module.getDocument(customer, AppConstants.USER_DOCUMENT_NAME);
+		AbstractPersistentBean adminUserBean = CORE.getPersistence()
+				.retrieve(userDocument, user.getId());
+		String contactEmailBinding = BindUtil.createCompoundBinding(AppConstants.CONTACT_ATTRIBUTE_NAME,
+				AppConstants.EMAIL1_ATTRIBUTE_NAME);
+		String contactNameBinding = BindUtil.createCompoundBinding(AppConstants.CONTACT_ATTRIBUTE_NAME,
+				AppConstants.NAME_ATTRIBUTE_NAME);
+		String userEmail = (String) BindUtil.get(adminUserBean, contactEmailBinding);
+		String userName = (String) BindUtil.get(adminUserBean, contactNameBinding);
+
+		// Check if the ip address changed since last login and if so log the event
+		String lastIpAddress = (String) BindUtil.get(previousLoginRecord, IP_ADDRESS);
+		if (userIPAddress != lastIpAddress) {
+			SecurityUtil.log("Change of IP Address from Last Login",
+					String.format(IP_CHANGE_LOG_MESSAGE, userName, lastIpAddress, userIPAddress));
+
+		}
+
+		// Check if the country has changed since the last login and if so send the user a warning message
+		String previousCountryCode = (String) BindUtil.get(previousLoginRecord, "countryCode");
+
+		if (countryCode != null && countryCode != previousCountryCode) {
+			// Get actual country name
+			Locale locale = new Locale("", countryCode);
+			String country = locale.getDisplayCountry();
+
+			Locale previousLocale = new Locale("", previousCountryCode);
+			String previousCountry = previousLocale.getDisplayCountry();
+
+			SecurityUtil.log("User Logged in from Different Country",
+					String.format(COUNTRY_CHANGE_LOG_MESSAGE, userName, previousCountry, country));
+
+			// Send email to user advising password reset
+			try {
+				EXT.sendMail(new Mail().from(UtilImpl.SMTP_SENDER)
+						.addTo(userEmail)
+						.subject("Security Alert: Unusual Login Activity Detected")
+						.body(String.format(UNUSUAL_LOGIN_EMAIL_BODY, userName, country, UtilImpl.SUPPORT_EMAIL_ADDRESS)));
```

Comment:
This should use CommunicationUtil, see some of the other security cards Simeon did. This allows the email to be changed by the customer, instead of hard coded in Skyve, so we don't need a new Skyve version to update the email.

Reply:
I have updated the code but I need to find a way to run the send email in a separate thread like a job, but I'm not sure if I can use jobs in skyve-ext

Reply:
Hey @Malcolm-G-Biz - I think you can move all this code to UserLoginBizlet. See my comment at end.